### DISCUSSION TITLE
useFieldArray - revalidation after dynamic fields added

### DISCUSSION THREAD

**Discussion by lucascebertin** [react-hook-form/react-hook-form#4955] [Q&A] [answered]
I would like to know if there is something that could trigger revalidation after dynamic fields added with defaultValues in screen 

Well, chances that I'm doing something wrong here but I have something like this:
https://codesandbox.io/s/sad-shirley-nk12c?file=/src/ComplexForm.js

See, when clicked on the radio button `Item2` you'll get a list item (with `useFieldArray`) and the submit button remains `disabled` as soon as you click/enter in some textbox and just leaves it then submit will be enabled.

1) Is this expectation correct by my side?
2) If yes, there is a way to achieve that? 

Sorry, I forgot to mention all versions being used here:
```json
"dependencies": {
    "@emotion/react": "^11.1.5",
    "@emotion/styled": "^11.3.0",
    "@hookform/resolvers": "1.3.0",
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.1.0",
    "@testing-library/user-event": "^12.1.10",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-hook-form": "^6.15.4",
    "react-scripts": "4.0.3",
    "web-vitals": "^1.0.1",
    "yup": "^0.32.9"
  }
```

**Comment by lucascebertin** [comment]
Alright, I did something here that looks like what I need... any comments, suggestions, anything, is always welcome
https://codesandbox.io/s/romantic-fermat-wgl3h

**Reply by lucascebertin** [reply]
Yeah, nevermind, same problem in a different place...
Now, when enter in third radio button the validation occurs but when leaves it's uses the previous data which is still valid until the complete removal/render...

**Comment by bluebill1049** [comment]
> I would like to know if there is something that could trigger revalidation after dynamic fields added with defaultValues in screen

can you provide more detail on that? so you want to trigger validation when a new input gets append?

**Comment by lucascebertin** [comment]
Yes, of course.
I'll try to explain... the first codesandbox will show you some yup validation schemas.
```js
  const itemSchema = yup.object().shape({
    name: yup.string().required()
  });

  const schema = yup.object().shape({
    radio: yup.number().required(),
    item: yup.array().of(itemSchema).required()
  });
```
I'm destructuring the isValid this way:
```js
  const methods = useForm({
    defaultValues,
    mode: "all",
    resolver: yupResolver(schema)
  });
  
  const {
    handleSubmit,
    formState: { isValid }
  } = methods;
``` 
My submit button is disabled like this way:
```js
 <Input type="submit" value="Submit form" disabled={!isValid} />
 ```

I expect the `isValid` returning true only when the radio `Item2` is checked.

What I got is:
1) Form is loaded, radio button item 0 is auto selected:
![image](https://user-images.githubusercontent.com/6990857/116170213-bedb6f80-a6dc-11eb-8fe2-586d90b9c536.png)

2) Check radio button Item2, 3 textboxes are appended to the form and the isValid return false
![image](https://user-images.githubusercontent.com/6990857/116170236-cf8be580-a6dc-11eb-99c1-7e65749c1402.png)

3) Check radio button Item1, 3 textboxes are removed from the form and for some reason the isValid return true
![image](https://user-images.githubusercontent.com/6990857/116170347-fea25700-a6dc-11eb-99b9-93a96cad6e79.png)

4) Check radio button Item0, there is no textboxes anymore into the form and now the isValid returns false
![image](https://user-images.githubusercontent.com/6990857/116170279-e29eb580-a6dc-11eb-8528-4893e0b4f97b.png)

**Reply by bluebill1049** [reply]
<img width="579" alt="Screen Shot 2021-04-27 at 11 38 35 am" src="https://user-images.githubusercontent.com/10513364/116171943-1adddc80-a74d-11eb-9854-a70b30c84971.png">

Do you have to run `trigger` here as well?

**Reply by lucascebertin** [reply]
Don't know if run the `trigger` is the answer because I don't need exactly the messages being shown but the `isValid` changing.
Makes any sense?

**Reply by lucascebertin** [reply]
We just get rid of using the `isValid`.
Now, we are just `watch()`ing all form values and manually understanding `when` the submit button should be enabled.

**Comment by leapful** [comment, marked as answer]
@lucascebertin @bluebill1049 

The root cause of this issue is from mount/unmount behavior of React.

When selecting the last radio "Item 2", "onChange" is triggered with validation but at that time, all 3 child inputs are not rendered => isValid = false. 

After selecting another "Item 0" or "Item 1", the same handling of "onChange" happens and at that time, all 3 child inputs are not removed yet => isValid = true.

I would suggest a workaround to trigger validation after everything is rendered successfully by using "setTimeout":

https://codesandbox.io/s/green-framework-tz2ur?file=/src/ComplexForm.js

**Reply by lucascebertin** [reply]
Yeah, we found the `setTimeout` approach but we just decided to don't use that because is a little bit hard to explain for everyone involved the reasons behind it.
Is that a "scenario" that could be handled inside `react-hook-form`?

**Reply by lucascebertin** [reply]
BTW, @leapful, thanks for sharing your findings and the approach, really appreciate that!

**Reply by bluebill1049** [reply]
Thanks very much @leapful any chance you can try the latest build? maybe this is getting resolved? we don't unregister at useEffect now, hopefully, this makes a difference.

**Reply by lucascebertin** [reply]
Sorry, what build number you mean?
I think I'm already using the latest for v6. (6.15.7)

**Reply by bluebill1049** [reply]
maybe you can try with v7.

**Reply by lucascebertin** [reply]
Maybe I didn't understand correctly but v7 works with react v16.13.1?

**Reply by bluebill1049** [reply]
yes.

**Reply by lucascebertin** [reply]
Great, I'll try, thanks
